reduce overhead in calls to multi-dimensional FFTs.

grlee77 opened the pull request:
This PR refactors some of the axis handling logic into a `_prep_fftn_axes` helper function which can be memoized, reducing FFT call overhead. Overall this PR also results in a small net reduction in lines of code. The determination of the output array dtype was also memoized here.

Prior to this PR, on a somewhat slow older PC, a call to `_default_fft_func` (called by all non-1d functions to determine whether to use 1D or nD planning) took approximately 40 microseconds. After the refactoring here, it takes ~1 microsecond. Prior to this PR, the overhead could be a substantial fraction of the total `fftn` time for smaller FFTs (e.g. 512 x 512) when using pre-planning + plan context manager.

The changes here are within `cupy.fft.fft` and so will also benefit `cupyx.scipy.fftpack` and upcoming `cupyx.scipy.fft` as well.

grlee77 commented:
I have changed this to WIP while I rebase and fix some failing test cases from the recently merged `scipy.fft` module.

grlee77 commented:
Okay, now all new scipy.fft test cases pass as well. I do see a couple of failures related to `irfftn` as reported in #2569, but those also occur on current master and may be caused by a bug in cuFFT itself in CUDA 10.1 update 2.

toslunar commented on cupy/fft/fft.py:
```diff
@@ -1,34 +1,38 @@
 
-def _output_dtype(a, value_type):
+@cupy.util.memoize()
+def _output_dtype(dtype, value_type):
```
This `memoize` looks great.

grlee77 commented on cupy/fft/fft.py:
```diff
@@ -175,6 +180,45 @@ def _fft(a, s, axes, norm, direction, value_type='C2C', overwrite_x=False,
     return a
 
 
+@cupy.util.memoize()
```
I agree that not allowing axes as a list is problematic.

I will just remove `memoize` from this function.

leofang replied:
Hi @toslunar @grlee77 I was tracing back to the commit a1d7e60 and found this discussion. 

I think the concern 1 (unbounded combinations) is probably too much, because other CuPy functions that already use `@memoize()` can also have many many possible combinations, and yet we're using it without noticeable performance issue. The key here is the actual combinations that one can run into in each runtime session is likely small. 

For concern 2, what if we make `axes` a tuple before calling this function? It's an invariant in this function anyway.

grlee77 commented:
@toslunar:

I have addressed the comments and added two additional tests cases corresponding to:
1.) fftn/ifftn with `axes=()`
2.) fftn/ifftn with `axes` as a list of integers rather than a tuple

The behavior of `cupy.fft.fftn/ifftn` for `axes=()` has been modified to match NumPy's behavior (in 1.17, I did not check older versions). In this case where no axes are transformed, the input array should be returned unmodified, but prior to this PR, CuPy would cast the dtype to complex.

toslunar commented on cupy/fft/fft.py:
```diff
@@ -175,6 +185,47 @@ def _fft(a, s, axes, norm, direction, value_type='C2C', overwrite_x=False,
+            not all([(axes_sorted[n + 1] - axes_sorted[n]) == 1
+                     for n in range(len(axes_sorted) - 1)])):
+        return False
+    return True
```
- Do you want to keep the error message for `if len(fft_axes) < 1 or len(fft_axes) > 3`?
- How about the following?
```python
    return (
        0 < len(axes_sorted) <= 3
        and (0 in axes_sorted or (ndim - 1) in axes_sorted)
        and all([
            (axes_sorted[n + 1] - axes_sorted[n]) == 1
            for n in range(len(axes_sorted) - 1)]))
```

grlee77 replied:
This form looks better, thanks. As far as the other error message you mention, I think we should remove it and just mention 1-3 axes in the case that `_nd_plan_is_possible` returns False.

toslunar commented on tests/cupy_tests/fft_tests/test_fft.py:
```diff
@@ -259,6 +261,10 @@ def test_fftn(self, xp, dtype, enable_nd):
         a = testing.shaped_random(self.shape, xp, dtype)
         out = xp.fft.fftn(a, s=self.s, axes=self.axes, norm=self.norm)
 
+        if self.axes is ():
```
```suggestion
        if self.axes == ():
```

toslunar commented on tests/cupy_tests/fft_tests/test_fft.py:
```diff
@@ -259,6 +261,10 @@ def test_fftn(self, xp, dtype, enable_nd):
         out = xp.fft.fftn(a, s=self.s, axes=self.axes, norm=self.norm)
 
+        if self.axes is ():
+            assert out is a
```
OK to test this undocumented behavior.

toslunar commented:
Could you also fix the two lines with `is ()` in the tests? (https://github.com/cupy/cupy/pull/2746#discussion_r356518811)

grlee77 commented:
> Could you also fix the two lines with is () in the tests? (#2746 (comment))

Thanks for the reminder. I overlooked that one before. It should be fixed now.

toslunar commented:
Jenkins, test this please.

grlee77 commented:
Not sure what is causing the failures here, but the errors don't appear related to the changes in this PR.

toslunar commented:
Jenkins, test this please.

toslunar commented on cupy/fft/fft.py:
```diff
@@ -203,30 +252,14 @@ def _get_cufft_plan_nd(shape, fft_type, axes=None, order='C'):
         fft_axes = tuple(range(ndim))
     else:
-        if np.isscalar(axes):
-            axes = (axes, )
```
Note: OK to remove `axes = (axes, )` because there has been the following incompatibility.
```
>>> scipy.fftpack.fftn(numpy.ones((2, 3)), axes=0)
array([[2.+0.j, 2.+0.j, 2.+0.j],
       [0.+0.j, 0.+0.j, 0.+0.j]])
>>> cupyx.scipy.fftpack.fftn(cupy.ones((2, 3)), axes=0)
TypeError: 'int' object is not iterable
```

The behavior is different between numpy and scipy.
- `numpy.fft.fftn` rejects non-sequence `axes`.
- `scipy.fftpack.fftn` and `scipy.fft.fftn` accept non-sequence `axes`.
